VCSM dmabuf import

6by9 opened the pull request:
Firmware and userland PRs to come.

6by9 committed:
VCSM: Specify pointers in IPC msgs as uint32_t

Pointers were specified as void* which will fail
rather badly if the kernel is switched to 64 bit.
GPU wants 32 bit addresses, so use uint32_t and
cast whereever necessary.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
VCSM: New option to import a DMABUF for VPU use

Takes a dmabuf, and then calls over to the VPU to wrap
it into a suitable handle.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: code-style: Fix comment indentation

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Convert to loading via device tree.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

popcornmix commented:
No objections here. A quick test builds and runs existing code that uses vcsm.
@pelwell?

pelwell commented on drivers/char/broadcom/vc_sm/vmcs_sm.c:
```diff
@@ -1749,8 +1751,8 @@ static int vc_sm_ioctl_lock(struct SM_PRIV_DATA_T *private,
-		} else if (status != 0 ||
-			   (status == 0 && result.res_mem == NULL)) {
+		} else if (status ||
+			   (!status && !(void *)result.res_mem)) {
```
Isn't this cast pointless?

6by9 replied:
Yes. Done.
Too many passes through checkpatch.
res_mem being converted to u32 meant that comparison of u32 to NULL (ptr) became a warning. Then it prefers !(ptr) to ptr == NULL hence why we ended up here.

Just looking through and there are a bundle of ```if (X || (!X && Y)```, which either have redundant casts but can also be reduced to ```if (X || Y)```. I'll nuke those too.

pelwell commented on drivers/char/broadcom/vc_sm/vc_vchi_sm.h:
```diff
@@ -19,63 +19,73 @@
-/* Terminates the shared memory service.
-*/
+/*
+ *Terminates the shared memory service.
```
Missing space.

6by9 replied:
Done

pelwell commented on drivers/char/broadcom/vc_sm/vmcs_sm.c:
```diff
@@ -3104,15 +3096,24 @@ static void vc_sm_connected_init(void)
+	sm_state = kzalloc(sizeof(*sm_state), GFP_KERNEL);
+	if (!sm_state)
+		return -ENOMEM;
+	pr_err("vc-sm: pdev is %p. dev is %p\n", pdev, &pdev->dev);
```
pr_info at most.

6by9 replied:
Removed. Left over debugging.

pelwell commented on drivers/char/broadcom/vc_sm/vc_sm_defs.h:
```diff
@@ -165,6 +181,41 @@ typedef struct {
 } VC_SM_FREE_ALL_T;
 
+/* Request to import memory (HOST->VC) */
+struct VC_SM_IMPORT {
```
Linux doesn't use CAPS for structure names.

6by9 replied:
I retained caps for consistency with the rest of the file. This file is also shared with the firmware.
So either we go through all the code on both sides and adopt the Linux scheme (and ignore the firmware scheme), or we retain the current style.

Then again I didn't retain full consistency as I dropped the typedefs in favour of named structs.

What's the preference here? A clean up set of patches to go the Linux way?

pelwell replied:
I would choose to not introduce new rule violations rather than maintain consistency.

6by9 replied:
OK, will do.
I'll do a follow up clean up set of patches to try and make vc-sm as a whole friendlier to mainline.

pelwell commented on drivers/char/broadcom/vc_sm/vc_vchi_sm.h:
```diff
@@ -89,4 +89,11 @@ int vc_vchi_sm_walk_alloc(VC_VCHI_SM_HANDLE_T handle);
 			VC_SM_ACTION_CLEAN_T *action_clean);
 
+/*
+ * Import a contiguous block of memory and wrap into in a GPU MEM_HANDLE_T.
```
"it in a"?

6by9 replied:
Done

pelwell commented on drivers/char/broadcom/vc_sm/vmcs_sm.c:
```diff
@@ -500,7 +510,9 @@ static void vmcs_sm_remove_map(struct SM_STATE_T *state,
 static int vc_sm_global_state_show(struct seq_file *s, void *v)
 {
 	struct sm_mmap *map = NULL;
+	struct SM_RESOURCE_T *resource = NULL;
```
See CAPS comment.

pelwell commented on drivers/char/broadcom/vc_sm/vmcs_sm.c:
```diff
@@ -2056,6 +2110,138 @@ static int vc_sm_ioctl_unlock(struct SM_PRIV_DATA_T *private,
+	import.size = sg_dma_len(sgt->sgl);
+	import.allocator = current->tgid;
+
+	if (*ioparam->name) {
```
Lose the braces.

6by9 replied:
Checkpatch disagrees due to braces on the else clause.
CHECK: braces {} should be used on all arms of this statement
#396: FILE: drivers/char/broadcom/vc_sm/vmcs_sm.c:2164:
+	if (*ioparam->name)
[...]
+	else {
[...]

pelwell replied:
But the else clause doesn't need them either.

6by9 replied:
Ack. Think I'd looked at a different instance of almost identical syntax.

pelwell commented on drivers/char/broadcom/vc_sm/vmcs_sm.c:
```diff
@@ -2056,6 +2110,138 @@ static int vc_sm_ioctl_unlock(struct SM_PRIV_DATA_T *private,
+		private->restart_sys = -EINTR;
+		private->int_action = VC_SM_MSG_TYPE_IMPORT;
+		goto error;
+	} else if (status != 0 || (status == 0 && result.res_handle == 0)) {
```
`status || !result.res_handle`

6by9 replied:
Done

6by9 committed:
vcsm: Remove all typedefs from vc_sm_defs.h and calling code

Remove typedefs on the structures that make up the IPC
to the firmware for VCSM. Update all calling code appropriately.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Correct block comment style on licence headers.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix block comment formatting

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Remove typedef of VC_VCHI_SM_HANDLE_T

Replaced with struct sm_instance *.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: "unsigned int" preferred over "unsigned" fixed.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix blank lines after declarations.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix erroneous space in 'if' statement.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix lots of block quote formatting issues

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix spaces around operators.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix indentation of switch/case statement.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix use of S_IRUGO and use 0444 instead

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Correct "long unsigned int" to "unsigned long"

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix "struct vm_operations_struct should normally be const"

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix "void function return statements are not generally useful"

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Fix "Prefer kcalloc over kzalloc with multiply"

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Remove VC_SM_LOCK_CACHE_MODE_T typedef.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 commented:
Original 4 commits amended as requested.

16 followups fixing various things that checkpatch complains about. It still leaves a load of line > 80 chars / too many tab indents in vmcs_sm.c, but those can wait.

6by9 commented:
Please don't merge yet - I've still to fully test all of these! It compiles, but hasn't been run, and I've got to pull the IPC header over and make the firmware mods still.

pelwell commented:
That was close. ;-)

6by9 commented:
Oops, sorry.
Dmabuf import needs the firmware merge request to be able to do anything useful anyway.

6by9 committed:
vcsm: Replace SM_RESOURCE_T with sm_resource_t

Follow the Linux kernel coding style.

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Replace SM_STATS_T with sm_stats_t

Follows the kernel coding style

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Replace SM_STATE_T with sm_state_t

Follow the Linux kernel coding style

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Replace SM_PRIV_DATA_T with sm_priv_data_t

Follows the Linux kernel coding style

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 committed:
vcsm: Replace SM_PDE_T with sm_pde_t

Follows the Linux kernel coding style

Signed-off-by: Dave Stevenson <dave.stevenson@raspberrypi.org>

6by9 commented:
OK, I'm happy. 
Branch rebased and 6 extra patches added, but they're just switching structure names from all caps to lower case to follow the kernel coding style.
checkpatch with --strict still has 55 warnings and 228 checks, but those really will wait.

One thing that may want addressing is the definition of enum vc_sm_alloc_type_t
```
enum vc_sm_alloc_type_t {
	VC_SM_ALLOC_CACHED,
	VC_SM_ALLOC_NON_CACHED,
};
```
This is used in the IPC message structures, but doesn't guarantee the size of the enum. It's not shown up in practice, but I think that is only because the following fields need 4 byte alignment. Do you agree that needs fixing, or not care?

@popcornmix - do you want to rerun your tests before this is merged? Mine have all been fine, and inspection shows no change in behaviour beyond the new feature.

popcornmix commented:
@6by9 might be worth adding a dummy enum entry
```VC_SM_ALLOC_FORCE_ENUM_32BIT = 0x80000000```
to make sure a smaller enum type is not used.
(or change to uint32_t).

6by9 commented:
@popcornmix That was my thought too. I was a little surprised that it wasn't there already.

I suspect I'll do another round of vcsm cleanups some time soonish (1-2 weeks) as it will be required upstream if they're wanting to use MMAL efficiently. At least if we've done the first round of cleanups then they might whinge less about it. I'll roll the enum update into the next set of cleanups as it isn't a blocking issue here due to where vc_sm_alloc_type_t is placed in the structures.
